packages/ember-routing/lib/location/auto_location.js

```diff
@@ -73,285 +86,178 @@ export default {
-    this._location.replace(this._getOrigin() + path);
-  },
-
-  /**
-    @since 1.5.1
-    @private
-    @method _getRootURL
-  */
-  _getRootURL: function () {
-    return this.rootURL;
-  },
+    return this.container.lookup('location:' + implementation);
+  }),
 
-  /**
-    @private
+  initState: delegateToConcreteImplementation('initState'),
+  getURL: delegateToConcreteImplementation('getURL'),
+  setURL: delegateToConcreteImplementation('setURL'),
+  replaceURL: delegateToConcreteImplementation('replaceURL'),
+  onUpdateURL: delegateToConcreteImplementation('onUpdateURL'),
+  formatURL: delegateToConcreteImplementation('formatURL'),
 
-    Returns the current `location.pathname`, normalized for IE inconsistencies.
+  willDestroy: function() {
+    var concreteImplementation = get(this, 'concreteImplementation');
 
-    @method _getPath
-  */
-  _getPath: function () {
-    var pathname = this._location.pathname;
-    // Various versions of IE/Opera don't always return a leading slash
-    if (pathname.charAt(0) !== '/') {
-      pathname = '/' + pathname;
+    if (concreteImplementation) {
+      concreteImplementation.destroy();
     }
+  }
+});
 
-    return pathname;
-  },
-
-  /**
-    @private
-
-    Returns normalized location.hash as an alias to Ember.Location._getHash
-
-    @since 1.5.1
-    @method _getHash
-  */
-  _getHash: EmberLocation._getHash,
-
-  /**
-    @private
-
-    Returns location.search
-
-    @since 1.5.1
-    @method _getQuery
-  */
-  _getQuery: function () {
-    return this._location.search;
-  },
-
-  /**
-    @private
+function delegateToConcreteImplementation(methodName) {
+  return function() {
+    var concreteImplementation = get(this, 'concreteImplementation');
+    concreteImplementation[methodName].apply(concreteImplementation, arguments);
+  };
+}
 
-    Returns the full pathname including query and hash
+/**
+  Given the browser's `location`, `history` and `userAgent`, and a configured
+  root URL, this function detects whether the browser supports the [History
+  API](https://developer.mozilla.org/en-US/docs/Web/API/History) and returns a
+  string representing the Location object to use based on its determination.
 
-    @method _getFullPath
-  */
-  _getFullPath: function () {
-    return this._getPath() + this._getQuery() + this._getHash();
-  },
+  For example, if the page loads in an evergreen browser, this function would
+  return the string "history", meaning the history API and thus HistoryLocation
+  should be used. If the page is loaded in IE8, it will return the string
+  "hash," indicating that the History API should be simulated by manipulating the
+  hash portion of the location.
 
-  /**
-    @private
+*/
 
-    Returns the current path as it should appear for HistoryLocation supported
-    browsers. This may very well differ from the real current path (e.g. if it
-    starts off as a hashed URL)
+function detectImplementation(options) {
+  var location = options.location,
```

Comment:
> @stefanpenner How does that rule affect destructuring assignment, which is what I would do if I had ES6?
simply expands obvious structures, like above. That is why i said, this scenario like doesn't matter, just a good opportunity for to trick you into giving feedback on the style guide ;)

```js
let [a,b,c] = array;
let ({d,e,f}) = obj;
````

were it does matter is

```js
var a = doSomething();
var b = doSomethingElse();
var c = doSomethingVeryDifferent();
```